sys/kern/src/arch/rv64/helpers.rs

```diff
@@ -17,6 +20,34 @@ impl crate::atomic::AtomicExt for AtomicBool {
     }
 }
 
+// Use QEMU's memory-mapped sifive_test device to attempt
+// restart/poweroff.
+#[cfg(feature = "riscv-sifive_test-device")]
+mod sifive_test {
+    pub const RESET_ADDR: *mut u32 = 0x00100000 as *mut u32;
+    pub const REBOOT_VALUE: u32 = 0x00007777;
+    pub const _POWEROFF_VALUE: u32 = 0x00005555;
+    pub const _FAIL_VALUE: u32 = 0x00003333;
+}
```

Comment:
This shouldn't really be part of the arch rv64 code. We should find a way to add the abstraction layers for different chips to be used in the kernel.

Reply:
I was thinking along the same lines as your first option i.e. define a trait (or multiple traits) for chip. This trait can be implemented by each chip and we can have some safe defaults depending upon what features are available v/s not on a chip. This trait can be provided as an input to the kernel code either from app's main.rs or some other mechanism (compile time config selection). 

>  Alternatively, the chip's directory could just contain a rust file that contains any chip-specific info in a generic format that just gets added in by the build.rs.

I think this suffers from the same problem as adding things to chip.toml -- Not everything that we want to customize can be defined a generic format. It becomes much cleaner and simpler if we have a trait (or multiple traits) to define the functionality that would be required from the chip.

The reason why I say multiple traits is because I think it could provide a better organization if we divide the traits as per features or functionality rather than creating one giant chip trait. But, these are just my initial thoughts. We should explore and evaluate.

Reply:
> I think this suffers from the same problem as adding things to chip.toml -- Not everything that we want to customize can be defined a generic format. 

I don't think this is the best approach either. It's certainly simpler, and it means we don't need to mess with `start_kernel()` (as doing so would break ARM support without some intervention on our end), but traits provide a better abstraction layer.

> The reason why I say multiple traits is because I think it could provide a better organization if we divide the traits as per features or functionality rather than creating one giant chip trait. 

I think this is a reasonable approach.

> But, these are just my initial thoughts. We should explore and evaluate.

Agreed.

Reply:
> .. it means we don't need to mess with start_kernel() (as doing so would break ARM support without some intervention on our end),...

Yeah, we should be able to make ARM and RV work with differing requirements and optionality about chips.

Reply:
It looks like a fair amount of chip-specific (and board-specific) functionality is already defined and then exposed through the 'drv/' & 'idl/' mechanisms respectively (an example would be drv/user-leds, and idl/user-leds.idol).

It also looks like there have already been attempts to extend reset behaviour in the past (allowing a reset reason) and this functionality is provided by the supervisor (see 'request_reset' in jefe.idol).  Access to this is controlled through [tasks.jefe.config.allowed-callers] in app.toml
(See also 'set_reset_reason' in 'drv/lpc55-syscon/src/main.rs' to see some chip-specific reset reason code)

Rather than trying to work out the best way to get chip/board-specific code into the kernel, perhaps instead we should lean-into Hubris' driver-server model and instead seek to implement this sort of behaviour as 'drv/reset-api', 'idl/reset.idol' (or similar) and recognise that the initial inclusion of 'reset' in the kernel (and in arch-specific code) was a mistake.

Reply:
I don't see any reason it needs to be in the kernel.  I think having having a `reset-api` is a good idea.  When jefe receives a reset request, it can delegate it out to the target specific task.  This also provides a nice way to select your reset handler without needing kernel features for every target.

Reply:
I'm inclined to agree, but SBI calls need to be made from S-mode, right? If that's the case, I believe the kernel will need to have some involvement if SBI is going to be used to reset in S-mode firmware. 

This could be as simple as just keeping `kipc::system_restart()` exposed only on S-mode builds, but that would be putting arch-specific functionality into `kipc`.

Reply:
I agree that for any "device" related drivers, we shouldn't add that in the kernel. So, if the reset is being triggered by writing to some MMIO device, then yes it makes sense to add reset-api and use the idol interface to enable this.

However, I am thinking of cases where we might need any chip-related stuff that has to be done in the kernel e.g. some CSRs accessible only in M-mode or S-mode(wherever the kernel is running). There might be other cases as well that would require things to be done in the kernel. In such a case, we might have chip-specific implementation that the kernel needs and cannot delegate to a U-mode task. One example is what @FawazTirmizi pointed out above - mtimecmp and mtime. On RV, these are chip-specific implementations. Timer is currently tightly coupled with the kernel and so enabling  timer functionality requires some chip specific implementation in the kernel. We worked around that by adding to chip.toml and exposing as variables to the kernel. However, this might not always be possible or make organization really messy. Hence, my comment about adding chip-support to kernel (it doesn't apply just to this case of reset).

Reply:
> I'm inclined to agree, but SBI calls need to be made from S-mode, right?

Yes, SBI is the interface between S-mode and its execution environment. So, SBI calls are expected to come from S-mode.

> If that's the case, I believe the kernel will need to have some involvement if SBI is going to be used to reset in S-mode firmware.

Yep. There might not be a "lot" of chip related things there. But, that doesn't mean there won't be any. We would need support for vendor-specific SBI calls as well.

> This could be as simple as just keeping kipc::system_restart() exposed only on S-mode builds, but that would be putting arch-specific functionality into kipc.

We could probably create some logic around allowing what IPCs are allowed and this selection/policy could exist outside of the kipc library.

Reply:
I don't really like the idea of having allowed/disallowed kipc calls based on the platform.  I think having the kernel utilize platform specific functionality through a chip trait is better.  Then the kernel interface is not changing for different platforms.

Reply:
I think we should outline ideas outside this thread, https://docs.google.com/document/d/10YFJrb8jvJF13ZvE_Fc0GM-Vr0DNbxXEdAgqG-p9p9s/edit.

Reply:
Also, raised this [issue](http://go/ji/SW-1076) to track the discussion.